content/posts/how-to-evolve-application-cicd/index.md

```diff
@@ -0,0 +1,336 @@
+        - save_cache:
+            paths:
+              - ~/.gradle
+            key: v1-dependencies-{{ checksum "build.gradle" }}
+
+workflows:
+version: 2.1
+microservice:
+jobs:
+- assemble
+- unit-tests:
+requires:
+- assemble
+- integration-tests:
+requires:
+- assemble
+- system-tests:
+requires:
+- assemble
+
+````
+
+
+
+
+
+### Implementation with GitHub Action
+
+#### Terminology
+
+Workflow: In the official documentation a workflow is defined as: "A workflow is a configurable automated process that will execute one or more jobs.
+It is set up via a yaml file
+https://docs.github.com/fr/actions/using-workflows/about-workflows
+
+action: "Actions are individual tasks that you can combine to create jobs and customize your workflow."
+https://docs.github.com/fr/actions/creating-actions/about-custom-actions
+
+####  first connection
+
+GitHub Action is the integrated continuous integration tool of GitHUb.
+It is accessible from the Action button in a toolbar
+![github_action_menu.png](github_action_menu.png)
+
+At first access, GithUb Action displays a workflow proposal according to your project.
+In our case we have selected the java with gradle workflow.
+
+![github_action_gradle.png](github_action_gradle.png)
+
+Once selected our workflow, we are redirected on the editor proposing us a default file.
+It is possible to modify it and save it on a specific branch; In order to easily test the CICD without having to push on develop, it is possible to
+define the name of the working branch dna sla part push.
+
+![github_action_gradle_file.png](github_action_gradle_file.png)
+
+### Creation of jobs
+
+
+
+````yml
+on:
+  push:
+    branches:
+    - develop
+    - ci_* 
+````
+
+
+To implement the 3 steps, defined above, we need to create the jobs:
+
+* assemble: to compile the code and créer le jar.
+* spotless: to check the format of new code.
+* unit-tests: to execute the unit tests.
+* integration-tests: to execute the integration tests.
+* system-tests: to execute the system tests.
+
+the code below show an exemple to create a job
+
+
+````yml
+jobs:
+  assemble:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - uses: actions/setup-java@v3
+      with:
+        java-version: '17'
+        distribution: 'temurin'
+        cache: gradle
+    - run: ./gradlew assemble --no-daemon
+
+````
+
+
+In the above code :
+
+* we create the assemble job which is based on ubuntu-latest
+* we use two actions to check the code and configure Java
+* we run the gradle assemble job
+
+
+
+
+## Summary
+The first objective of this article was to test 3 cicd engines (GitLab, CircleCi and GitHub Action), implementing the same CI.
+While trying to implement the cicd on these engines, I quickly encountered some limitations for the first two:
+
+* on GitLab you can't push the result of the cicd to GitHub
+* on CircleCi the total size limit of the artefacts and their retention time were not compatible with our objectives.
```

Comment:
Can you say what are those limits?
Also, I think that nowhere in the article the objective of not spending a penny is undeclared.

Reply:
Add in the foreword the objective of not spending a penny.

Reply:
@dcremonini  done